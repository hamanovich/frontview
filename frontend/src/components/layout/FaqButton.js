import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import FontAwesome from 'react-fontawesome';

const FaqBtn = styled(Link)`
  position: fixed;
  right: -0.375rem;
  top: calc(50% - 3rem);
  font-size: 3rem;
`;

const FaqButton = () => (
  <FaqBtn to="/faq" className="btn btn-danger btn-lg">
    <FontAwesome name="question-circle-o" />
  </FaqBtn>
);

export default FaqButton;
