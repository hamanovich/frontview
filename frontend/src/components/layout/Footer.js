import React from 'react';
import styled from 'styled-components';

const Foot = styled.footer`
  position: absolute;
  bottom: 0;
  width: 100%;
  height: 60px;
  padding: 10px 0;
  background-color: #1a242f;
  text-align: center;
  color: #fff;
`;

const Footer = () => (
  <Foot>
    &copy; Front View <br /> 2017-2018
  </Foot>
);

export default Footer;
