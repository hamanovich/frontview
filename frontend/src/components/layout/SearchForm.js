import React from 'react';
import { func } from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import FontAwesome from 'react-fontawesome';

import Navbar from 'react-bootstrap/lib/Navbar';
import Form from 'react-bootstrap/lib/Form';
import Button from 'react-bootstrap/lib/Button';

import validate from '../../validations/search';

import { TextField } from '../formElements';

export const SearchForm = ({ handleSubmit, onSearch }) => (
  <Navbar.Form pullLeft>
    <Form onSubmit={handleSubmit(onSearch)} noValidate>
      <Field
        component={TextField}
        type="search"
        htmlFor="search"
        name="search"
        placeholder="Search question"
        feedback={false}
        errorsVisible={false}
      />{' '}
      <Button type="submit">
        <FontAwesome name="search" />
      </Button>
    </Form>
  </Navbar.Form>
);

SearchForm.propTypes = {
  handleSubmit: func.isRequired,
  onSearch: func.isRequired,
};

export default reduxForm({
  form: 'SearchForm',
  validate,
})(SearchForm);
