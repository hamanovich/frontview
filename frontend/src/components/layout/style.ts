import styled from 'styled-components';

import Navbar from 'react-bootstrap/Navbar';
import Form from 'react-bootstrap/Form';
import Image from 'react-bootstrap/Image';

export const Menu = styled(Navbar)`
  border-radius: 0;
`;

export const MediaImage = styled(Image)`
  width: 25px;
  height: 25px;
  max-width: 25px;
  margin: -10px 5px -10px 0;
`;

export const SForm = styled(Form)`
  .form-control.is-invalid,
  .form-control.is-valid {
    background-image: none;
    padding-right: 0.75rem;
  }

  .form-group {
    margin-bottom: 0;
    padding-right: 1rem;
  }

  @media (max-width: 767px) {
    display: flex;
    width: 100%;
    padding-top: 1rem;

    .form-group {
      flex-grow: 1;

      input {
        width: 100%;
      }
    }
  }
`;
