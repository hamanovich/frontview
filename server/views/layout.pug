doctype html
html(lang='en')
  head
    title= `${title}`
    meta(charset="UTF-8")
    meta(name="viewport" content="width=device-width, initial-scale=1")
    meta(http-equiv="X-UA-Compatible" content="ie=edge")
    link(rel="stylesheet" href="/styles/bootstrap.css")
    link(rel="shortcut icon" href="/favicon.ico" type="image/x-icon")
    link(rel="icon" href="/favicon.ico" type="image/x-icon")
    base(href="/")
  body

  .content
      block content
        p This is the default!

  block scripts
    script(src="bundle.js")