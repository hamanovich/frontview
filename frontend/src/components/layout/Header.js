import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { IndexLinkContainer, LinkContainer } from 'react-router-bootstrap';
import { connect } from 'react-redux';
import FontAwesome from 'react-fontawesome';
import styled from 'styled-components';

import Navbar from 'react-bootstrap/lib/Navbar';
import Nav from 'react-bootstrap/lib/Nav';
import NavItem from 'react-bootstrap/lib/NavItem';
import NavDropdown from 'react-bootstrap/lib/NavDropdown';
import MenuItem from 'react-bootstrap/lib/MenuItem';
import Image from 'react-bootstrap/lib/Image';

import { logout } from '../../actions/auth';
import { getUser } from '../../actions/signup';
import { getSearchedQuestions } from '../../actions/questions';
import { addFlashMessage } from '../../actions/flash';

import SearchForm from './SearchForm';

import { UserType } from '../../propTypes';

const Menu = styled(Navbar)`
  border-radius: 0;
`;

const MediaImage = styled(Image)`
  width: 25px;
  height: 25px;
  max-width: 25px;
  margin: -10px 5px -10px 0;
`;

const { shape, func, bool } = PropTypes;

class Header extends Component {
  static propTypes = {
    auth: shape({
      isAuthenticated: bool.isRequired,
      user: UserType.isRequired,
    }).isRequired,
    getSearchedQuestions: func.isRequired,
    getUser: func.isRequired,
    addFlashMessage: func.isRequired,
    logout: func.isRequired,
  };

  static contextTypes = {
    router: shape({
      history: shape({
        push: func.isRequired,
      }).isRequired,
    }).isRequired,
  };

  componentDidMount() {
    const { getUser, auth } = this.props;
    if (auth.isAuthenticated) {
      getUser(auth.user.username);
    }
  }

  onSearch = values => {
    const { getSearchedQuestions, addFlashMessage } = this.props;
    const { history } = this.context.router;

    if (!values.search) return;

    getSearchedQuestions(values.search).then(res => {
      if (!res.length) {
        addFlashMessage({
          type: 'warn',
          text: `Nothing found by search = ${values.search}`,
        });

        return;
      }

      history.push(`/questions/search?q=${values.search}`);
    });
  };

  render() {
    const { auth, logout, getSearchedQuestions, addFlashMessage } = this.props;
    const userPick = (
      <span>
        <MediaImage src={auth.user.gravatar} circle />
        {auth.user.username}
      </span>
    );
    const userLinks = (
      <Nav pullRight>
        <NavDropdown title="Menu" id="menu-dropdown">
          <LinkContainer to="/interview">
            <MenuItem>
              <FontAwesome name="id-badge" /> Interview
            </MenuItem>
          </LinkContainer>
          <MenuItem divider />
          <IndexLinkContainer to="/questions">
            <MenuItem>Show all questions</MenuItem>
          </IndexLinkContainer>
          <LinkContainer to="/questions/top">
            <MenuItem>
              <FontAwesome name="exclamation" /> Top 10
            </MenuItem>
          </LinkContainer>
          <LinkContainer to="/questions/internet">
            <MenuItem>
              <FontAwesome name="internet-explorer" /> From Internet
            </MenuItem>
          </LinkContainer>
          <MenuItem divider />
          <LinkContainer to="/questions/level">
            <MenuItem>
              <FontAwesome name="line-chart" /> By Levels
            </MenuItem>
          </LinkContainer>
          <LinkContainer to="/questions/skill">
            <MenuItem>
              <FontAwesome name="star-half-o" /> By Skills
            </MenuItem>
          </LinkContainer>
          <LinkContainer to="/questions/practice">
            <MenuItem>
              <FontAwesome name="keyboard-o" /> By Practice
            </MenuItem>
          </LinkContainer>
          <MenuItem divider />
          <LinkContainer to="/questions/add">
            <MenuItem>
              <FontAwesome name="question-circle-o" /> Add new
            </MenuItem>
          </LinkContainer>
        </NavDropdown>
        <NavDropdown title={userPick} id="account-dropdown">
          <IndexLinkContainer to="/me">
            <MenuItem>
              <FontAwesome name="user" /> Account
            </MenuItem>
          </IndexLinkContainer>
          <LinkContainer to="/me/edit">
            <MenuItem>
              <FontAwesome name="pencil-square-o" /> Edit profile
            </MenuItem>
          </LinkContainer>
          <MenuItem divider />
          <MenuItem onClick={logout}>
            <FontAwesome name="lock" /> Logout
          </MenuItem>
        </NavDropdown>
      </Nav>
    );

    const guestLinks = (
      <Nav pullRight>
        <LinkContainer to="/signup">
          <NavItem>
            <FontAwesome name="user-plus" /> Sign Up
          </NavItem>
        </LinkContainer>
        <LinkContainer to="/login">
          <NavItem>
            <FontAwesome name="user-circle" /> Login
          </NavItem>
        </LinkContainer>
      </Nav>
    );

    return (
      <Menu>
        <Navbar.Header>
          <Navbar.Brand>
            <Link to="/">Frontview /</Link>
          </Navbar.Brand>
          <Navbar.Toggle />
        </Navbar.Header>
        <Navbar.Collapse>
          <SearchForm
            getSearchedQuestions={getSearchedQuestions}
            addFlashMessage={addFlashMessage}
            onSearch={this.onSearch}
          />
          {auth.isAuthenticated ? userLinks : guestLinks}
        </Navbar.Collapse>
      </Menu>
    );
  }
}

const mapStateToProps = state => ({ auth: state.auth });

export default connect(
  mapStateToProps,
  {
    logout,
    getUser,
    getSearchedQuestions,
    addFlashMessage,
  },
  null,
  { pure: false },
)(Header);
