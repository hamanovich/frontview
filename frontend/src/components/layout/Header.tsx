import React, { Component } from 'react';
import { shape, func } from 'prop-types';
import { Link } from 'react-router-dom';
import { IndexLinkContainer, LinkContainer } from 'react-router-bootstrap';
import { connect } from 'react-redux';
import FontAwesome from 'react-fontawesome';

import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';
import NavDropdown from 'react-bootstrap/NavDropdown';

import { logout, getUser } from '../../actions/auth';
import { getSearchedQuestions } from '../../actions/questions';
import { addFlashMessage } from '../../actions/flash';
import SearchForm from './SearchForm';
import { Auth } from '../../propTypes/UserType';

import { Menu } from './style';

type HeaderProps = {
  getSearchedQuestions: (query: string) => Promise<any>;
  getUser: (identifier: string) => Promise<{}>;
  addFlashMessage: (payload: {
    type: string;
    text: string;
  }) => { type: string; payload: { type: string; text: string } };
  logout: () => void;
  auth: Auth;
};

type searchValues = {
  search: string;
};

export class Header extends Component<HeaderProps, {}> {
  static contextTypes = {
    router: shape({
      history: shape({
        push: func.isRequired,
      }).isRequired,
    }).isRequired,
  };

  componentDidMount() {
    const { getUser, auth, addFlashMessage } = this.props;

    if (auth.isAuthenticated) {
      getUser(auth.user.username).catch(() =>
        addFlashMessage({
          type: 'error',
          text:
            '__500: Server-side error__. Please check your internet connection',
        }),
      );
    }
  }

  private onSearch = (values: searchValues) => {
    const { getSearchedQuestions, addFlashMessage } = this.props;
    const { history } = this.context.router;

    if (!values.search) return;

    getSearchedQuestions(values.search)
      .then((res: { length: number }) => {
        if (!res.length) {
          addFlashMessage({
            type: 'warn',
            text: `Nothing found by search '__${values.search}__'`,
          });

          return;
        }

        history.push(`/questions/search?q=${values.search}`);
      })
      .catch(() =>
        addFlashMessage({
          type: 'error',
          text:
            '__500: Server-side error__. Please check your internet connection',
        }),
      );
  };

  render() {
    const { auth, logout } = this.props;
    const userPick = auth.user.username;
    // Known issue: https://github.com/react-bootstrap/react-bootstrap/issues/3534
    // const userPick = (
    //   <Fragment>
    //     <MediaImage src={auth.user.gravatar} roundedCircle />
    //     {auth.user.username}
    //   </Fragment>
    // );
    const userLinks = (
      <Nav>
        <NavDropdown title="Menu" id="menu-dropdown">
          <LinkContainer to="/interview">
            <NavDropdown.Item>
              <FontAwesome name="id-badge" /> Interview (in progress)
            </NavDropdown.Item>
          </LinkContainer>
          <NavDropdown.Divider />
          <IndexLinkContainer to="/questions">
            <NavDropdown.Item>Show all questions</NavDropdown.Item>
          </IndexLinkContainer>
          <LinkContainer to="/questions/top">
            <NavDropdown.Item>
              <FontAwesome name="exclamation" /> Top 10
            </NavDropdown.Item>
          </LinkContainer>
          <LinkContainer to="/questions/internet">
            <NavDropdown.Item>
              <FontAwesome name="internet-explorer" /> From Internet
            </NavDropdown.Item>
          </LinkContainer>
          <NavDropdown.Divider />
          <LinkContainer to="/questions/level">
            <NavDropdown.Item>
              <FontAwesome name="line-chart" /> By Levels
            </NavDropdown.Item>
          </LinkContainer>
          <LinkContainer to="/questions/skill">
            <NavDropdown.Item>
              <FontAwesome name="star-half-o" /> By Skills
            </NavDropdown.Item>
          </LinkContainer>
          <LinkContainer to="/questions/practice">
            <NavDropdown.Item>
              <FontAwesome name="keyboard-o" /> By Type
            </NavDropdown.Item>
          </LinkContainer>
          <NavDropdown.Divider />
          <LinkContainer to="/questions/add">
            <NavDropdown.Item>
              <FontAwesome name="question-circle-o" /> Add new
            </NavDropdown.Item>
          </LinkContainer>
        </NavDropdown>
        <NavDropdown title={userPick} id="account-dropdown">
          <IndexLinkContainer to="/me">
            <NavDropdown.Item>
              <FontAwesome name="user" /> Account
            </NavDropdown.Item>
          </IndexLinkContainer>
          <LinkContainer to="/me/edit">
            <NavDropdown.Item>
              <FontAwesome name="pencil-square-o" /> Edit profile
            </NavDropdown.Item>
          </LinkContainer>
          <NavDropdown.Divider />
          <NavDropdown.Item onClick={logout}>
            <FontAwesome name="lock" /> Logout
          </NavDropdown.Item>
        </NavDropdown>
      </Nav>
    );

    const guestLinks = (
      <Nav>
        <LinkContainer to="/signup">
          <Nav.Link>
            <FontAwesome name="user-plus" /> Sign Up
          </Nav.Link>
        </LinkContainer>
        <LinkContainer to="/login">
          <Nav.Link>
            <FontAwesome name="user-circle" /> Login
          </Nav.Link>
        </LinkContainer>
      </Nav>
    );

    return (
      <Menu variant="dark" bg="dark" expand="md">
        <Navbar.Brand>
          <Link to="/">Frontview /</Link>
        </Navbar.Brand>
        <Navbar.Toggle />
        <Navbar.Collapse>
          <SearchForm onSearch={this.onSearch} />
          <Nav className="mr-auto">
            <Navbar.Text>or</Navbar.Text>
            <LinkContainer to="/questions">
              <Nav.Link>
                Check All <FontAwesome name="question-circle" size="lg" />
              </Nav.Link>
            </LinkContainer>
          </Nav>
          {auth.isAuthenticated ? userLinks : guestLinks}
        </Navbar.Collapse>
      </Menu>
    );
  }
}

const mapStateToProps = (state: { auth: Auth }) => ({ auth: state.auth });

export default connect(
  mapStateToProps,
  {
    logout,
    getUser,
    getSearchedQuestions,
    addFlashMessage,
  },
  null,
  { pure: false },
)(Header);
