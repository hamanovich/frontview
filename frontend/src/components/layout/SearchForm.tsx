import React, { FC } from 'react';
import { Field, reduxForm, InjectedFormProps } from 'redux-form';
import FontAwesome from 'react-fontawesome';

import Button from 'react-bootstrap/Button';

import validate from '../../validations/search';

import { TextField } from '../formElements';

import { SForm } from './style';

type SearchFormProps = {
  onSearch: any;
};

export const SearchForm: FC<
  SearchFormProps & InjectedFormProps<{}, SearchFormProps>
> = ({ handleSubmit, onSearch }) => (
  <SForm
    onSubmit={handleSubmit(onSearch)}
    noValidate
    className="mr-2 form-inline">
    <Field
      component={TextField}
      type="search"
      htmlFor="search"
      name="search"
      placeholder="Search question"
      className="mr-sm-2"
      feedback={false}
      errorsVisible={false}
    />{' '}
    <Button type="submit">
      <FontAwesome name="search" />
    </Button>
  </SForm>
);

export default reduxForm<{}, SearchFormProps>({
  form: 'SearchForm',
  validate,
})(SearchForm);
