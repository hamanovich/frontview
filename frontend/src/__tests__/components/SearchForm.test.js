import React from 'react';
import { shallow } from 'enzyme';

import { SearchForm } from '../../components/layout/SearchForm';

describe('<SearchForm/>', () => {
  it('renders <SearchForm /> component', () => {
    const props = {
      handleSubmit: jest.fn(),
      onSearch: jest.fn(),
    };
    const component = shallow(<SearchForm {...props} />);

    expect(component).toMatchSnapshot();
  });
});
