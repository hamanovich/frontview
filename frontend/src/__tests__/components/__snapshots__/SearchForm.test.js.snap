// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm/> renders <SearchForm /> component 1`] = `
<Styled(Form)
  as="form"
  className="mr-2 form-inline"
  inline={false}
  noValidate={true}
>
  <Field
    className="mr-sm-2"
    component={[Function]}
    errorsVisible={false}
    feedback={false}
    htmlFor="search"
    name="search"
    placeholder="Search question"
    type="search"
  />
   
  <Button
    active={false}
    disabled={false}
    type="submit"
    variant="primary"
  >
    <FontAwesome
      name="search"
    />
  </Button>
</Styled(Form)>
`;
